exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/StatisticsAggBatch.java

```diff
@@ -0,0 +1,256 @@
+ *       "cnt"        : BIGINT - statscount(cnt)
+ *    "nonnullstatcount" : MAP
+ *       "region_id"  : BIGINT - nonnullstatcount(region_id)
+ *       "sales_city" : BIGINT - nonnullstatcount(sales_city)
+ *       "cnt"        : BIGINT - nonnullstatcount(cnt)
+ *   .... another map for next stats function ....
+ */
+public class StatisticsAggBatch extends StreamingAggBatch {
+  private List<String> functions;
+  private int schema = 0;
+
+  public StatisticsAggBatch(StatisticsAggregate popConfig, RecordBatch incoming,
+      FragmentContext context) throws OutOfMemoryException {
+    super(popConfig, incoming, context);
+    this.functions = popConfig.getFunctions();
+  }
+
+  private void createKeyColumn(String name, LogicalExpression expr, List<LogicalExpression> keyExprs,
+      List<TypedFieldId> keyOutputIds) throws SchemaChangeException {
+    ErrorCollector collector = new ErrorCollectorImpl();
+
+    LogicalExpression mle = ExpressionTreeMaterializer.materialize(expr, incoming, collector,
+        context.getFunctionRegistry());
+
+    MaterializedField outputField = MaterializedField.create(name, mle.getMajorType());
+    ValueVector vector = TypeHelper.getNewVector(outputField, oContext.getAllocator());
+
+    keyExprs.add(mle);
+    keyOutputIds.add(container.add(vector));
+
+    if (collector.hasErrors()) {
+      throw new SchemaChangeException("Failure while materializing expression. "
+          + collector.toErrorString());
+    }
+  }
+
+  private void createNestedKeyColumn(MapVector parent, String name, LogicalExpression expr,
+      List<LogicalExpression> keyExprs, List<TypedFieldId> keyOutputIds)
+          throws SchemaChangeException {
+    ErrorCollector collector = new ErrorCollectorImpl();
+
+    LogicalExpression mle = ExpressionTreeMaterializer.materialize(expr, incoming, collector,
+        context.getFunctionRegistry());
+
+    Class<? extends ValueVector> vvc =
+        TypeHelper.getValueVectorClass(mle.getMajorType().getMinorType(),
+            mle.getMajorType().getMode());
+
+    ValueVector vv = parent.addOrGet(name, mle.getMajorType(), vvc);
+
+    TypedFieldId pfid = container.getValueVectorId(SchemaPath.getSimplePath(parent.getField().getPath()));
+    assert pfid.getFieldIds().length == 1;
+    TypedFieldId.Builder builder = TypedFieldId.newBuilder();
+    builder.addId(pfid.getFieldIds()[0]);
+    TypedFieldId id =
+        FieldIdUtil.getFieldIdIfMatches(parent, builder, true,
+            SchemaPath.getSimplePath(vv.getField().getPath()).getRootSegment());
+
+    keyExprs.add(mle);
+    keyOutputIds.add(id);
+
+    if (collector.hasErrors()) {
+      throw new SchemaChangeException("Failure while materializing expression. "
+          + collector.toErrorString());
+    }
+  }
+
+  private void addMapVector(String name, MapVector parent, LogicalExpression expr,
+      List<LogicalExpression> valueExprs) throws SchemaChangeException {
+    ErrorCollector collector = new ErrorCollectorImpl();
```

Comment:
There is quite a bit of duplication across these three methods. While reuse by copy & paste has a long and storied tradition, it is actually easier to maintain code if we can move the repeated bits into functions that are called where needed. Also means us poor reviewers don't have to review the same code multiple times looking for subtle differences...